docs/tutorial_1/1_beam_introduction.ipynb

```diff
@@ -1,876 +1,1681 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@dataclass\n",
+    "class MovieTitle:\n",
+    "    movie_id: int\n",
+    "    year: int\n",
+    "    name: str"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Python Type Safety and Custom objects\n",
+    "\n",
+    "The Apache Beam SDK for Python uses type hints during pipeline construction and runtime to try to emulate the correctness guarantees achieved by true static typing. Additionally, using type hints lays some groundwork that allows the backend service to perform efficient type deduction and registration of `Coder` objects.\n",
+    "\n",
+    "Since in this tutorial we want to handle our own custom **data object**, we also need to implement our custom (_yet simple_) `beam.coders.Coder` to allow for a correct object _serialisation_.\n",
+    "\n",
+    "**Note**: This step is particularly crucial during the execution (_in parallel_) of the Pipeline, to avoid `apache_beam` to fallback to the default `Serializer` (i.e. `Coder`) based in `pickle`.\n",
+    "More information on **Python Type Safety** in Apache Beam pipelines is available in the official [documentation](https://beam.apache.org/documentation/sdks/python-type-safety/)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MovieViewCoder(beam.coders.Coder):\n",
+    "    \"\"\"Beam Coder Serialiser for custom MovieView Data Object\"\"\"\n",
+    "    def encode(self, view):\n",
+    "        return (f\"{view.user_id}:{view.movie_id}:{view.rating}:{str(view.date.date())}\").encode(\"utf-8\")\n",
+    "\n",
+    "    def decode(self, s):\n",
+    "        return MovieView(*s.decode(\"utf-8\").split(':'))\n",
+    "\n",
+    "    def is_deterministic(self):\n",
+    "        return True\n",
+    "    \n",
+    "\n",
+    "class MovieTitleCoder(beam.coders.Coder):\n",
+    "    \"\"\"Beam Coder Serialiser for custom MovieTitle Data Object\"\"\"\n",
+    "    def encode(self, title):\n",
+    "        # using a separation character which will unlikely appear in titles\n",
+    "        return (f\"{title.movie_id}--{title.year}--{title.name}\").encode(\"utf-8\")\n",
+    "    \n",
+    "    def decode(self, s):\n",
+    "        return MovieTitle(*s.decode(\"utf-8\").split('--'))\n",
+    "    \n",
+    "    def is_deterministic(self):\n",
+    "        return True\n",
+    "\n",
+    "\n",
+    "# Once defined, we can register the new Coders\n",
+    "beam.coders.registry.register_coder(MovieView, MovieViewCoder)\n",
+    "beam.coders.registry.register_coder(MovieTitle, MovieTitleCoder)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Workaround `multiprocessing` issues with namespace"
```

Comment:
Please add optional, and that it's not effective for Google Colab to the name of this subsection and hide the content of this subsection by default